doc: add docs for dealing with special characters in ref names (#9577)
github/docs: content/get-started/using-git/dealing-with-special-characters-in-branch-and-tag-names.md

---
title: Dealing with special characters in branch and tag names
intro: 'Git is very permissive about what characters are allowed in branch and tag names. When using Git from a command-line shell, you may need to deal with special characters by using escaping or quoting.'
versions:
  fpt: '*'
  ghes: '*'
  ghae: '*'
shortTitle: Special characters in names
---

> Most repositories use simple branch names like `main` or `update-icons`. Likewise, tag names usually follow a basic format, such as a version number like `v1.2.3`. Both branch and tag names may also use the path separator (`/`) for structure (for example, `area/item`, or `level-1/level-2/level-3`). Apart from a handful of exceptions &mdash; such as not starting or ending a name with a slash, or having consecutive slashes in the name &mdash; Git places [very few restrictions](https://git-scm.com/docs/git-check-ref-format) on what characters may be used.

Reviewer: Small accessibility suggestion: can we expand the link text from “very few restrictions” to a larger part of the sentence, something like “very few restrictions on what characters may be used in branch and tag names.” Ideally it would make sense in isolation, e.g. when using a screen reader in links mode.

(see [WCAG 2.1, success criterion 2.4.4: Link Purpose (In Context)][2.4.4] and [WCAG 2.1, success criterion 2.4.9: Link Purpose (Link Only)][2.4.9])

[2.4.4]: https://www.w3.org/WAI/WCAG21/Understanding/link-purpose-in-context.html
[2.4.9]: https://www.w3.org/WAI/WCAG21/Understanding/link-purpose-link-only.html

Author: Good call. Thanks @georgebrock!

* In the [Bash](https://www.gnu.org/software/bash/) shell, enclosing a string of characters in single quotes preserves the literal value of the characters within the single quotes (but note, you cannot include a single quote inside a single-quoted string).
* [Zsh](https://www.zsh.org/) behaves similarly to Bash, in that single quotes turn off the special behavior of most characters inside them (although, note that this behavior is configurable via the `RC_QUOTES` option).
* [PowerShell](https://microsoft.com/powershell) likewise treats characters verbatim when inside single quotes.

> As such, most branch and tag names with special characters can be handled by wrapping the name in single quotes. The main exception to this rule is when the branch name itself contains a single quote. For edge cases like this, it is best to consult the official documentation provided with your shell. Starting points for some common shells include:

Reviewer: What do you think about moving the first sentence of this paragraph before the list? It seems like the list could use a little more introduction to connect it to the idea of quoting or escaping appropriately in the prior paragraph.

Something like:

> Most branch and tag names with special characters can be quoted by wrapping the name in single quotes:
>
> - In the Bash shell, …
>
> The main exception to this rule is …

Author: Sounds good @georgebrock. 👍

---
title: Dealing with special characters in branch and tag names
intro: 'Git is very permissive about what characters are allowed in branch and tag names. When using Git from a command-line shell, you may need to escape or quote special characters.'
versions:
  fpt: '*'
  ghes: '*'
  ghae: '*'
shortTitle: Special characters in names
---

## About branch and tag names

Most repositories use simple branch names, such as `main` or `update-icons`. Tag names also usually follow a basic format, such as a version number like `v1.2.3`. Both branch names and tag names may also use the path separator (`/`) for structure, for example `area/item` or `level-1/level-2/level-3`. Other than some exceptions &mdash; such as not starting or ending a name with a slash, or having consecutive slashes in the name &mdash; Git has very few restrictions on what characters may be used in branch and tag names. For more information, see "[git-check-ref-format](https://git-scm.com/docs/git-check-ref-format)" in the Git documentation.

> ## Why escaping special characters is necessary

Reviewer 2: ```suggestion
## Why you need to escape special characters
```

> When using a CLI, you might have situations where a branch or tag name contains special characters that have a special meaning for your shell environment. These characters must be quoted or escaped for you to use them safely in a Git command, otherwise running the command may have unintended effects.

Reviewer 2: ```suggestion
When using a CLI, you might have situations where a branch or tag name contains special characters that have a special meaning for your shell environment. To use these characters safely in a Git command, they must be quoted or escaped, otherwise the command may have unintended effects.
```

> For example, the `$` character is used by many shells to refer to a variable. Most shells would interpret a valid branch name like `hello-$USER` as equivalent to the word "hello", followed by a hyphen, followed by the current value of the `USER` variable, rather than the literal string `hello-$USER`. If a branch name includes the `$` character, then we need to stop the shell from expanding it as a variable reference. Similarly, if a branch name contains a semi-colon (`;`), most shells interpret it as a command separator, so we need to quote or escape it appropriately.

Reviewer 2: ```suggestion
For example, the `$` character is used by many shells to refer to a variable. Most shells would interpret a valid branch name like `hello-$USER` as equivalent to the word "hello", followed by a hyphen, followed by the current value of the `USER` variable, rather than the literal string `hello-$USER`. If a branch name includes the `$` character, then the shell must be stopped from expanding it as a variable reference. Similarly, if a branch name contains a semi-colon (`;`), most shells interpret it as a command separator, so it needs to be quoted or escaped.
```

## How to escape special characters in branch and tag names

Most branch and tag names with special characters can be handled by including the name in single quotes, for example `'hello-$USER'`.

* In the [Bash](https://www.gnu.org/software/bash/) shell, enclosing a string of characters in single quotes preserves the literal value of the characters within the single quotes.
* [Zsh](https://www.zsh.org/) behaves similar to Bash, however this behavior is configurable using the `RC_QUOTES` option.
* [PowerShell](https://microsoft.com/powershell) also treats characters literally when inside single quotes.

For these shells, the main exception is when the branch or tag name itself contains a single quote. In this case, you should consult the official documentation for your shell:

* [Bash documentation](https://www.gnu.org/software/bash/manual/)
* [Zsh documentation](https://zsh.sourceforge.io/Doc/)
* [Fish documentation](https://fishshell.com/docs/current/)
* [PowerShell documentation](https://docs.microsoft.com/en-gb/powershell/)

## Naming branches and tags

> If possible, create branch and tag names that don't contain special characters which must be escaped. A safe default set of characters to use for branch names and tag names is:

Reviewer 2: ```suggestion
If possible, create branch and tag names that don't contain special characters, as these would need to be escaped. A safe default set of characters to use for branch names and tag names is:
```
